README.md
[Private Geth PoA Ethereum Network Setup Tutorial](https://www.youtube.com/playlist?list=PLkM0MH7Grb25poKEiId5pEQg-OzLQRNM4)


node_01/history
eth.accounts
var tx = { from: '0x08a427de46773b907a9c406f81948aa22d2d3260', to: '0x528EA91FA33F1C194E176929F22Adda8e4687d82', value: 50000000000000000000000000000000 };
eth.sendTransaction(tx);
web3.personal
eth.accounts
eth.getBalance("0x08a427de46773b907a9c406f81948aa22d2d3260")
eth.getBalance("0x528EA91FA33F1C194E176929F22Adda8e4687d82")
var tx = { from: '0x08a427de46773b907a9c406f81948aa22d2d3260', to: '0x528EA91FA33F1C194E176929F22Adda8e4687d82', value: 50000000000000000000000000000000 };
tx
eth.sendTransaction(tx);
var tx = { from: '0x08a427de46773b907a9c406f81948aa22d2d3260', to: '0x528EA91FA33F1C194E176929F22Adda8e4687d82', value: 50000000000000000000000000000000 };
eth.sendTransaction(tx);
eth.accounts
eth.getBalance("0x08a427de46773b907a9c406f81948aa22d2d3260")
eth.getBalance("0x528EA91FA33F1C194E176929F22Adda8e4687d82")
eth.accounts
eth.getBalance("0x528EA91FA33F1C194E176929F22Adda8e4687d82")
eth.getBalance("0x08a427de46773b907a9c406f81948aa22d2d3260")
var tx = { from: '0x08a427de46773b907a9c406f81948aa22d2d3260', to: '0x528EA91FA33F1C194E176929F22Adda8e4687d82', value: 50000000000000000000000000000000 };
eth.sendTransaction(tx);
eth.accounts
eth.getBalance("0x08a427de46773b907a9c406f81948aa22d2d3260")
var tx = { from: '0x08a427de46773b907a9c406f81948aa22d2d3260', to: '0x528EA91FA33F1C194E176929F22Adda8e4687d82', value: 50000000000000000000000000000000 };
eth.sendTransaction(tx);
var tx = { from: '0x08a427de46773b907a9c406f81948aa22d2d3260', to: '0x528EA91FA33F1C194E176929F22Adda8e4687d82', value: 5000000000000000000000000000000 };
eth.sendTransaction(tx);
eth.accounts

eth.py
# built-in packages import
import argparse
from decimal import Decimal

# 3rd party packages import
from web3 import Web3
from web3.middleware import geth_poa_middleware

# where should i use geth_poa_middleware?


# create a parser object
parser = argparse.ArgumentParser(description="Script Description")

# add arguments
parser.add_argument("command", help="Command to execute", choices=[
                    "getBalance", "sendEth", "getAccounts"])
parser.add_argument("-a", "--ownAddress", help="Owner's Address")
parser.add_argument("-t", "--toAddress", help="Other's Address")
parser.add_argument("-n", "--amount", help="Amount of ETH to send")

# parse the arguments from command line
args = parser.parse_args()

# get the arguments
command = args.command
ownAddress = args.ownAddress
toAddress = args.toAddress
amount = args.amount

# connect to the Ethereum node
w3 = Web3(Web3.HTTPProvider("http://localhost:8545"))
w32 = Web3(Web3.HTTPProvider("http://localhost:8546"))

w3.middleware_onion.inject(geth_poa_middleware, layer=0)
# w32.middleware_onion.inject(geth_poa_middleware, layer=0)

# check if my private blockchain network is connected


def isConnected():
    return w3.is_connected()

# check existing accounts in the node


def getAccounts():
    if not isConnected():
        # error with message not connected to ethereum node
        raise (Exception("Not connected to Ethereum node"))

    # List available Ethereum accounts
    node1accounts = w3.eth.accounts
    node2accounts = w32.eth.accounts
    allAccounts = node1accounts + node2accounts
    if len(allAccounts) > 0:
        return allAccounts
    else:
        return []


def getBalance(address):
    if not isConnected():
        # error with message not connected to ethereum node
        raise (Exception("Not connected to Ethereum node"))

    # Convert the address to a checksum address
    address_to_check = w3.to_checksum_address(address)
    if not address_to_check:
        print("Address is not valid!!")
        return None

    # Check the balance of the address
    balance_wei = w3.eth.get_balance(address_to_check)

    # Convert the balance from wei to ether
    balance_eth = w3.from_wei(balance_wei, 'ether')

    return balance_eth


def sendEth(fromAddress, toAddress, amount):
    if not isConnected():
        # error with message not connected to ethereum node
        raise (Exception("Not connected to Ethereum node"))

    amount_wei = w3.to_wei(amount, 'ether')

    # check if enough balance available
    if Decimal(getBalance(fromAddress)) < int(amount):
        print(f"Available balance: {Decimal(getBalance(fromAddress))} ETH")
        print(f"Requested amount: {amount} ETH")
        print(f"Not enough balance to send {amount} ETH from {fromAddress}")
        return False

     # Get the current base fee
    base_fee = w3.eth.gas_price

    # Calculate maxFeePerGas and maxPriorityFeePerGas
    max_fee_per_gas = base_fee
    max_priority_fee_per_gas = w3.to_wei('1', 'gwei')

    # Send the ETH
    tx_hash = w3.eth.send_transaction({
        'from': fromAddress,
        'to': toAddress,
        'value': amount_wei,
        'maxFeePerGas': max_fee_per_gas,
        'maxPriorityFeePerGas': max_priority_fee_per_gas,
    })

    # Get the transaction receipt
    tx_receipt = w3.eth.wait_for_transaction_receipt(tx_hash)

    # Check if the transaction was successful
    if tx_receipt['status'] == 1:
        print(f"Transaction successful. Hash: {tx_hash}")
    else:
        print(f"Transaction failed. Hash: {tx_hash}")
        return False

    return True

# check if an account number exists in the blockchain


def isAccount(accountNumber):
    if not isConnected():
        # error with message not connected to ethereum node
        raise (Exception("Not connected to Ethereum node"))

    # Check if the account exists
    if accountNumber in getAccounts():
        return True
    else:
        return False


# create account
# eb4394e531cb8812e118be301f15a72fc05fb58c99cff6f44bdbe4d044f74b52
# 0xb42d4ad620Cd7C86Fd3dE15dD65cD0FA64bC3fF3


if command == "getAccounts":
    print(getAccounts())
elif command == "getBalance":
    if not isAccount(ownAddress):
        print("Address is not valid")
    else:
        if ownAddress:
            balance = getBalance(ownAddress)
            if balance:
                print(f"Balance of {ownAddress} is {balance} ETH")
            else:
                print(f"Address {ownAddress} not found")
        else:
            print("Missing argument for -getBalance.")
elif command == "sendEth":
    if ownAddress and toAddress and amount:
        sendEth(ownAddress, toAddress, amount)
    else:
        print("Missing argument for -sendEth.")
else:
    print("Invalid command. Please use -getBalance or -sendEth.")

commands.sh
geth account new --datadir node_01
geth account new --datadir node_02

geth init --datadir node_01 genesis.json
geth init --datadir node_02 genesis.json

bootnode -genkey boot.key
bootnode -nodekey boot.key -addr :30305

clef --keystore ./node_01/keystore --configdir ./clef --chainid 123454321 --suppress-bootwarn init
clef --keystore ./node_01/keystore --configdir ./clef --chainid 123454321 --suppress-bootwarn setpw 0x08a427de46773b907a9c406f81948aa22d2d3260
clef --keystore ./node_01/keystore --configdir ./clef --chainid 123454321 --suppress-bootwarn --nousb

geth --datadir node_01 --port 30306 --bootnodes enode://a52f108d96da0318eca414177f19b08688e007a8bf7142b7b62f7a47771230155809f055012e33197d318a69d748c48f521e1ed88d2b54bcfe073241305f56bf@127.0.0.1:0?discport=30301 --networkid 123454321 --unlock 0x22e6eEb50319650DABE208c2e00eDf9D80b571C4 --password node_01/password.txt --authrpc.port 8551 --mine --miner.etherbase 0x1585917171590cA54D324D047c7B9c5ac6a6a589
geth --datadir node_02 --port 30307 --bootnodes enode://a52f108d96da0318eca414177f19b08688e007a8bf7142b7b62f7a47771230155809f055012e33197d318a69d748c48f521e1ed88d2b54bcfe073241305f56bf@127.0.0.1:0?discport=30301 --networkid 123454321 --unlock 0x9c901034B699864908792C03e720823268ED4122 --password node_02/password.txt --authrpc.port 8552

# this one doesnt work
# geth --datadir node_01 --port 30306 --bootnodes enode://a52f108d96da0318eca414177f19b08688e007a8bf7142b7b62f7a47771230155809f055012e33197d318a69d748c48f521e1ed88d2b54bcfe073241305f56bf@127.0.0.1:0?discport=30301 --networkid 123454321 --unlock 0x22e6eEb50319650DABE208c2e00eDf9D80b571C4 --password node_01/password.txt --rpc --rpcaddr '0.0.0.0' --rpcport 8501 --rpcapi 'personal,db,eth,net,web3,txpool,miner' --rpccorsdomain "*" --mine --miner.etherbase 0x22e6eEb50319650DABE208c2e00eDf9D80b571C4 --gasprice '1'

geth --datadir ./node_01 --unlock "0x08A427De46773b907a9C406F81948Aa22D2D3260" --password node_01/password.txt --networkid 123454321 --mine --miner.etherbase "0x08A427De46773b907a9C406F81948Aa22D2D3260" --port 30306 --networkid 123454321 --rpc --rpcaddr '0.0.0.0' --rpcport 8501 --rpccorsdomain "*"
geth --datadir ./node_02 --unlock "0x2ee97631432aD774AAC16f73AC6892fae6D817db" --password node_02/password.txt --networkid 123454321

geth --datadir ./node_02 --signer ./clef/clef.ipc console

geth attach node_01/geth.ipc

# Geth Javascript Console
net.peerCount
admin.peers

eth.accounts
eth.getBalance(eth.accounts[0])

# send some Wei
eth.sendTransaction({
  to: '0x2ee97631432aD774AAC16f73AC6892fae6D817db',
  from: eth.accounts[0],
  value: 25
});

eth.sendTransaction({  to: '0x528EA91FA33F1C194E176929F22Adda8e4687d82',  from: eth.accounts[0],  value: 50000000000000000000000000000000 });
var tx = { from: '0x08a427de46773b907a9c406f81948aa22d2d3260', to: '0x528EA91FA33F1C194E176929F22Adda8e4687d82', value: 50000000000000000000000000000000 };
eth.sendTransaction(tx);

# check the transaction was successful by querying Node 2's account balance
eth.getBalance('0x2ee97631432aD774AAC16f73AC6892fae6D817db');

# doesnt work
# eth.getBalance('0x2ee97631432aD774AAC16f73AC6892fae6D817db').then(result => web3.utils.fromWei(result,"ether"))

web3.eth.getBalance('0x528EA91FA33F1C194E176929F22Adda8e4687d82', function (error, wei) {
    if (!error) {
        var balance = web3.utils.fromWei(wei, 'ether');
        console.log(balance + " ETH");
    }
});


# Hootum - hootum
geth --datadir /root/hootum/node1 --port 30306 --bootnodes enode://210425feb653676de7014b463324e7c62933740a590e55b2868a461d8c1b1463dd7d998d61f7ba1aeeee780d0c80789d526fbb78ed5276cc40a37895bf8ee462@127.0.0.1:0?discport=30305 --networkid 138129 --unlock 0x0a179cc93fd1bd008Db536Be2828941B8E43D682 --password /root/hootum/node1/password.txt --mine --miner.etherbase 0x0a179cc93fd1bd008Db536Be2828941B8E43D682 --http --http.addr 0.0.0.0 --http.vhosts=chain.hootum.io,localhost --allow-insecure-unlock --syncmode full
geth --datadir /root/hootum/node3 --port 30326 --bootnodes enode://210425feb653676de7014b463324e7c62933740a590e55b2868a461d8c1b1463dd7d998d61f7ba1aeeee780d0c80789d526fbb78ed5276cc40a37895bf8ee462@127.0.0.1:0?discport=30305 --networkid 138129 --unlock 0x813E669712eDb6c8916BaB58b3F5F933cB3F96De --password /root/hootum/node3/password.txt --mine --miner.etherbase 0x813E669712eDb6c8916BaB58b3F5F933cB3F96De --authrpc.port 8562 --http --http.port 8547 --http.addr 0.0.0.0 --http.vhosts=chain.hootum.io,localhost --allow-insecure-unlock --syncmode full


# Hootum - hootum-chain-02
geth --datadir /root/hootum-chain-02/node4 --port 30336 --bootnodes enode://210425feb653676de7014b463324e7c62933740a590e55b2868a461d8c1b1463dd7d998d61f7ba1aeeee780d0c80789d526fbb78ed5276cc40a37895bf8ee462@64.227.158.244:0?discport=30305 --networkid 138129 --unlock 0x0cC7e5D30867e6f0c2573eE4Fd9eACcd0B8742B3 --password /root/hootum-chain-02/node4/password.txt --mine --miner.etherbase 0x0cC7e5D30867e6f0c2573eE4Fd9eACcd0B8742B3 --authrpc.port 8563 --http --http.port 8548 --http.addr 0.0.0.0 --http.vhosts=chain.hootum.io --allow-insecure-unlock --syncmode full
geth --datadir /root/hootum-chain-02/node5 --port 30346 --bootnodes enode://210425feb653676de7014b463324e7c62933740a590e55b2868a461d8c1b1463dd7d998d61f7ba1aeeee780d0c80789d526fbb78ed5276cc40a37895bf8ee462@64.227.158.244:0?discport=30305 --networkid 138129 --unlock 0xFfffcC8579De94b29B6bf82E7552DDd5EFa9DAE5 --password /root/hootum-chain-02/node5/password.txt --mine --miner.etherbase 0xFfffcC8579De94b29B6bf82E7552DDd5EFa9DAE5 --authrpc.port 8564 --http --http.port 8549 --http.addr 0.0.0.0 --http.vhosts=chain.hootum.io --allow-insecure-unlock --syncmode full
